community/detectors/xwiki_cve_2024_21650/src/main/java/com/google/tsunami/plugins/detectors/rce/cve202421650/Cve202421650Detector.java

```diff
@@ -0,0 +1,233 @@
+      "rest/wikis/xwiki/spaces/XWiki/pages/" + REQUEST_USER_NAME + "/objects/XWiki.XWikiUsers/0";
+
+  private static final String REQUEST_CLEANUP_FIRST_NAME_REPLACEMENT = "Delete Me!";
+
+  private final Clock utcClock;
+  private final HttpClient httpClient;
+
+  private final PayloadGenerator payloadGenerator;
+
+  @Inject
+  Cve202421650Detector(
+      @UtcClock Clock utcClock, HttpClient httpClient, PayloadGenerator payloadGenerator) {
+    this.utcClock = checkNotNull(utcClock);
+    this.httpClient = checkNotNull(httpClient);
+    this.payloadGenerator = checkNotNull(payloadGenerator);
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("Cve202421650Detector starts detecting.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(NetworkServiceUtils::isWebService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    String targetUri =
+        NetworkServiceUtils.buildWebApplicationRootUrl(networkService) + REQUEST_PATH;
+
+    String targetCleanupUri =
+        NetworkServiceUtils.buildWebApplicationRootUrl(networkService) + REQUEST_CLEANUP_PATH;
+
+    PayloadGeneratorConfig config =
+        PayloadGeneratorConfig.newBuilder()
+            .setVulnerabilityType(PayloadGeneratorConfig.VulnerabilityType.BLIND_RCE)
+            .setInterpretationEnvironment(
+                PayloadGeneratorConfig.InterpretationEnvironment.LINUX_SHELL)
+            .setExecutionEnvironment(
+                PayloadGeneratorConfig.ExecutionEnvironment.EXEC_INTERPRETATION_ENVIRONMENT)
+            .build();
+
+    Payload payload = payloadGenerator.generate(config);
+    String cmd = payload.getPayload();
+
+    String token = "";
+
+    try {
+      HttpResponse response =
+          httpClient.send(get(targetUri).withEmptyHeaders().build(), networkService);
+      if (response.status().code() != HttpStatus.OK.code()) {
+        return false;
+      }
+
+      Matcher csrfTokenMatcher = CSRF_TOKEN_PATTERN.matcher(response.bodyString().orElse(""));
+      if (csrfTokenMatcher.find()) {
+        token = csrfTokenMatcher.group(1);
+      }
+
+      if (token.isEmpty()) {
+        return false;
+      }
+
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Unable to request '%s'.", targetUri);
+    }
+
+    String requestBody = REQUEST_POST_DATA;
+
+    String[] placeholders = {"{{PAYLOAD}}", "{{TOKEN}}"};
+
+    String[] replacements = {cmd, token};
+
+    for (int i = 0; i < placeholders.length; i++) {
+      requestBody = requestBody.replace(placeholders[i], replacements[i]);
+    }
```

Comment:
```suggestion
    String requestBody = REQUEST_POST_DATA
            .replace("{{PAYLOAD}}", cmd)
            .replace("{{TOKEN}}", token);
```

Since there are only two values to replace, I think it's easier to just do two replace operations instead of a loop.

Reply:
agree, applied